### ISSUE TITLE
TimeseriesGenerator for labeled time-series such as sensor data

### ISSUE THREAD

**Issue by mmalekzadeh** [tensorflow/tensorflow#33566]
**System information**
- TensorFlow version: v2.0.0
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
Currently,  `TimeseriesGenerator` gets `data` and `target` as input and return a generator that can be used for iterating over a time-series with a sliding window.
- If `row` is the current time-point,  then for each time-window of length `[row - self.length, ..., row`] , the  `target` is chosen as the`self.targets[row]`.  This means the `target` (label) for a time-window is chosen to be the last observed label.

Sometimes, this is not the right assumption. For example, consider this time-series and the related labels:
```python
data    = [a,a,a,a,a,a,a,a,a,a,a,b,b,b,b,b,b,b,b,b,b,b,b,b,c,c,c,c,c,c,c,c]
target =  [0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2]
```
If we choose `length=8` and `stride=4`, then it generates these windows:
```python
a,a,a,a,a,a,a,a >  0  (Good label!)
a,a,a,a,a,a,a,b >  1  (Bad label!)
a,a,a,b,b,b,b,b > 1 (Not Good label!)
b,b,b,b,b,b,b,b > 1 (Good label!)
b,b,b,b, c,c,c,c > 2 (Bad label!)
c,c,c,c,c,c,c,c, > 2 (Good label!)
```
So, as you see, if I use this method in training (e.g. for an activity recognition method based on accelerometer time-seires) then some windows will be labeled in a Not Good way!

**There should be a feature to choose if we want this behavior or not!**

There are two ideas:
1. to simply ignore these windows with an inconsistent label in the beginning and the end! I mean just jump over them until we get a window that label is same for all the time-points. like this:
```python
targets = np.array([self.targets[row] for row in rows if self.targets[row - self.length] == self.targets[row]])
```
2. Using the `mode` of the target window as the label. like this:
```python 
targets = np.array([ max(set(self.targets[row - self.length:row:self.sampling_rate] ), key=list.count) for row in rows])
```

**Will this change the current api? How?**
There are two ways to handle this:
1.  To add more options to the current `TimeseriesGenerator` that will change the current api for this class.
2. To add another class and call it something like "LabeledTimeseriesGenerator" that extends the `TimeseriesGenerator` and overrides the `__getitem__` function.

**Who will benefit with this feature?**
People who use Tensorflow for training a classifier on labeled time-series of mobile and wearable sensors or using financial data time-series.

**Comment by tanzhenyu** [comment]
Thanks for the write-up.
I'm not sure I understand the use case here -- this generator are mostly for RNN models, which we do need target to be the label as value of the last time step.
Are you using this for other things?

**Comment by mmalekzadeh** [comment]
> Thanks for the write-up.
> I'm not sure I understand the use case here -- this generator are mostly for RNN models, which we do need target to be the label as value of the last time step.
> What's `sensor_data` in this case and how do you wanna use it?

The use case explained here is mostly useful in other kinds of time series (such as mobile and wearable sensor data) or when time series are processed in a sliding window manner.

In such a situation, the label is the majority across all the labels in the current window, not the last one.

BTW, maybe it is much better to have a class extended from this for that purpose.

**Comment by rjurney** [comment]
@mmalekzadeh I'm confused. Why can't you just shift your target to be however many items ahead you want? Wouldn't that achieve what you want? You can do this in pandas, for example with [pandas.Series.shift(-lookout_period)](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.shift.html).

**Comment by rmothukuru** [comment]
@mmalekzadeh,
Sorry for the delayed response. The use that you provided is not completely clear. Confusing part in your use-case is that you are using only one label for a series of Inputs. Please refer the latest [Tutorial on Time Series](https://www.tensorflow.org/tutorials/structured_data/time_series) and let us know if you still need some additional features.

Thanks!

**Comment by Saduf2019** [comment]
@mmalekzadeh 
Please post this issue on [keras-team/keras repo.](https://github.com/keras-team/keras/issues)
To know more see;
https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999

**Comment by mmalekzadeh** [comment]
I agree that there are different interpretations of `target` for a time window.

What I was trying to suggest is that if we consider `length=N`, this means that we generate a window including `N` samples. 
We also have `N` targets associated with each of these `N` samples.

So, currently, this method considers the last target as the target of the whole window. 

But in some applications, users may need to choose the `mode` of the `N` targets.

This is not a critical feature though as it depends to the application.

**Comment by tilakrayal** [comment]
@mmalekzadeh ,
Is this still an issue?
Could you please update TensorFlow to the latest stable version v.2.6 and let us know if you are facing the same error. Thanks!
